src/gtc/dace/expansion/sdfg_builder.py

```diff
@@ -0,0 +1,401 @@
+# GTC Toolchain - GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2022, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+import dataclasses
+from dataclasses import dataclass
+from typing import Any, ChainMap, Dict, List, Optional, Set, Tuple
+
+import dace
+import dace.data
+import dace.library
+import dace.subsets
+
+import eve
+import gtc.common as common
+from eve import NodeVisitor
+from gtc import daceir as dcir
+from gtc.dace.expansion.tasklet_codegen import TaskletCodegen
+from gtc.dace.expansion.utils import get_dace_debuginfo
+from gtc.dace.utils import data_type_to_dace_typeclass, make_dace_subset
+
+
+class StencilComputationSDFGBuilder(NodeVisitor):
+    contexts = (eve.SymbolTableTrait.symtable_merger,)
+
+    @dataclass
+    class NodeContext:
+        input_node_and_conns: Dict[Optional[str], Tuple[dace.nodes.Node, Optional[str]]]
+        output_node_and_conns: Dict[Optional[str], Tuple[dace.nodes.Node, Optional[str]]]
+
+    @dataclass
+    class SDFGContext:
+        sdfg: dace.SDFG
+        state: dace.SDFGState
+        state_stack: List[dace.SDFGState] = dataclasses.field(default_factory=list)
+
+        def add_state(self):
+            old_state = self.state
+            state = self.sdfg.add_state()
+            for edge in self.sdfg.out_edges(old_state):
+                self.sdfg.remove_edge(edge)
+                self.sdfg.add_edge(
+                    state,
+                    edge.dst,
+                    edge.data,
+                )
+            self.sdfg.add_edge(old_state, state, dace.InterstateEdge())
+            self.state = state
+            return self
+
+        def add_loop(self, index_range: dcir.Range):
+
+            loop_state = self.sdfg.add_state()
+            after_state = self.sdfg.add_state()
```

Comment:
I'm not sure this is doing what you think it is. Both `loop_state` and `after_state` should be the same thing here, since it mutates the referenced `self`.